### ISSUE TITLE
The "_loopFinshed" in TimerService.volatile should be changed to "_loopFinished"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6657]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| kaizhang reported|
|-|-|
|Trac ID|trac#6657|
|Type|defect|
|Created|2013-08-01 13:23:59Z|
|Branch|https://github.com/twisted/twisted/tree/timerservice-typo-6657|

It's a typo. It makes pickling a `twisted.application.internet.TimerService` instance try to pickle `TimerService._loopingFinished`.

<details><summary>Searchable metadata</summary>

```
trac-id__6657 6657
type__defect defect
reporter__kaizhang kaizhang
priority__normal normal
milestone__ 
branch__branches_timerservice_typo_6657 branches/timerservice-typo-6657
branch_author__kaizhang kaizhang
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1375363439000000 1375363439000000
changetime__1377001553000000 1377001553000000
version__None None
owner__kaizhang kaizhang

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|kaizhang set status to `assigned`|
|-|-|

It should be changing "_loopFinshed" to "_loopFinished". Sorry about the typo in summary and description.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|kaizhang commented|
|-|-|

(In [39412]) Branching to 'timerservice-typo-6657'

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/timerservice-typo-6657)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst set owner to kaizhang|
|-|-|

As we discussed on IRC, please rewrite the tests so they test the underlying issue (unpickleability) rather than an implementation detail.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|kaizhang set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

This looks ready for review?

 1. I'd do `self.assertFalse(hasattr(timer, "_loop"))` rather than `assertRaises`. More direct.
 1. The news file is too low-level in its explanation; you should be saying something like "TimerService is now pickleable in all cases."

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> I'd do self.assertFalse(hasattr(timer, "_loop")) rather than assertRaises. More direct.

But use the good spelling of hasattr, which is:

```python
    nothing = object()
    value = getattr(timer, "_loop", nothing)
    self.assertNotIdentical(nothing, value)
```

Always avoid `hasattr`, which swallows exceptions and turns them into `False` (differing from all real-world attribute access behavior where the exception will actually get raised and break your program).

Or you can use `assertRaises` without the lambda: `self.assertRaises(AttributeError, getattr, timer, "_loop")`

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|kaizhang commented|
|-|-|

Replying to [itamar](#note_7):
> This looks ready for review?
> 
>  1. I'd do `self.assertFalse(hasattr(timer, "_loop"))` rather than `assertRaises`. More direct.
>  1. The news file is too low-level in its explanation; you should be saying something like "TimerService is now pickleable in all cases."

Yes, it's ready for review. I didn't submit this because there were two buildbots offline when I finished. I will update my code according to your comments then submit it for review.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|kaizhang removed owner<br>kaizhang set status to `new`|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/timerservice-typo-6657)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst set owner to kaizhang|
|-|-|

Looks good, please merge.

**Comment by twisted-trac** [comment]
|<img alt="kaizhang's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|kaizhang set status to `closed`|
|-|-|

(In [39644]) Merge timerservice-typo-6657: Fix typo in TimerService.volatile

Author: kaizhang
Reviewer: itamar
Fixes: [#6657](https://github.com/twisted/twisted/issues/6657)

The "_loopFinshed" in TimerService.volatile should be changed to "_loopFinished". It's a typo. It makes pickling a twisted.application.internet.TimerService instance try to pickle TimerService._loopingFinished.
